File: Code/rotate_amino_acid.py
############################################################################
# run_dihedral.py
#
# run_dihedral.py folder coord_file AA_name coord_num save_folder save_name
#
# Loads the gro file of a Valine residue and rotates the full phi, psi, chi space
# Calculates the energy due to atomic overlaps at each position and writes to a file
# Angles are sampled every 5 degree
#
# Input:
#   folder: path to folder with the coordinate file
#   coord_file: file name containing the coordinates. Coordinates must be in gromacs order but in Angstroms
#   AA_name: amino acid 3 letter abbreviation
#   coord_num: integer of which set of corodinates this is
#   save_folder: folder to save data
#   save_name: additional string to use in save file names
#
# Required files
#   coord_file: coordinates in Gro order, columns = x, y, z, in Angstroms
#   Val_bonds.txt: atom indexes (1 indexed) of all atomic bonds
#   Val_angles.txt: atom indexes of all bond angles
#   Val_radii.txt: Atomic radii (in Angstroms) for each atom
#   Val_Hbonds.txt: 1 indexed list of potential hydrogen bonding atoms
#
# Saves:
#   save_folder + 'Val_energy_' + save_name + '_' + coord_num + '.txt'
#       Columns:
#           1: phi
#           2: psi
#           3: chi 1
#           4: Energy
#
############################################################################

import numpy as np
import sys
from calcDihedral import calcDihedral
from rotate_DA import rotate_DA
from create_clash_list import create_clash_list
from rotate_CH3 import rotate_CH3
from rotate_end_CH3 import rotate_end_CH3

# Get arguments
folder = sys.argv[1]
coord_file = sys.argv[2]
AA_name = sys.argv[3]
coord_num = sys.argv[4]
save_folder = sys.argv[5]
save_name = sys.argv[6]

# Load coordinates
Coord = np.loadtxt(folder + coord_file)        # in  Angstroms

# Load list of bonds, angles and atomic radii
bonds = np.loadtxt('Val_bonds.txt')                     # 1 indexed list of bond indexes
angles = np.loadtxt('Val_angles.txt')                   # 1 indexed list of bond angle indexes
radii = np.loadtxt('Val_radii.txt')                     # Atomic radii in Angstroms
hbonds = np.loadtxt('Val_Hbond.txt', dtype=int)

n_atoms = Coord.shape[0]

# fix bonds and angles to 0 indexing
bonds = bonds - 1
angles = angles - 1
hbonds = hbonds - 1

# Create clash list
clash_list = create_clash_list(n_atoms, bonds, angles)
radii_sum = radii[clash_list[:, 0]] + radii[clash_list[:, 1]]

# modify H-bond radii
for i in range(0, hbonds.shape[0]):
    ind1 = np.isin(clash_list[:, 0], hbonds[i, 0])
    ind2 = np.isin(clash_list[:, 1], hbonds[i, 1])
    radii_sum[ind1 & ind2] = 1.5

# Get sum of radii for all possible atom pairs in clash list
radii_2 = radii_sum * radii_sum                                 # Leave as (radii_sum)^2 to improve math later

# Get index of peptide caps
prior_res_index = [0, 1, 2, 3, 4, 5]
next_res_index = np.arange(n_atoms - 6, n_atoms)


# Initialize dihedral arrays
phi_index = np.array([4, 6, 8, n_atoms - 8])
psi_index = np.array([6, 8, n_atoms - 8, next_res_index[0]])
chi1_index = np.array([6, 8, 10, 12])
CH3_1_index = np.array([8, 10, 12, 13])
CH3_2_index = np.array([8, 10, 16, 17])
end_CH3_1_index = np.array([6, 4, 1, 0])
end_CH3_2_index = np.array([20, 22, 24, 25])

# Get initial dihedral angle values
phi_init = calcDihedral(phi_index, Coord)
psi_init = calcDihedral(psi_index, Coord)
chi1_init = calcDihedral(chi1_index, Coord)
CH3_1_init = calcDihedral(CH3_1_index, Coord)
CH3_2_init = calcDihedral(CH3_2_index, Coord)
end_CH3_1_init = calcDihedral(end_CH3_1_index, Coord)
end_CH3_2_init = calcDihedral(end_CH3_2_index, Coord)
omega_1 = calcDihedral([1, 4, 6, 8], Coord)
omega_2 = calcDihedral([8, 20, 22, 24], Coord)

# initial values of angles being rotated have to be > 0? not sure why
if (phi_init < 0):
    phi_init = phi_init + 360
if (psi_init < 0):
    psi_init = psi_init + 360
if (chi1_init < 0):
    chi1_init = chi1_init + 360

# Create list of atoms to move
moveAtomID_phi = np.arange(phi_index[2] + 1, n_atoms)
moveAtomID_psi = np.arange(psi_index[2] + 1, n_atoms)
moveAtomID_chi1 = np.arange(chi1_index[2] + 1, n_atoms - 8)
moveAtomID_CH3_1 = np.arange(13, 16)
moveAtomID_CH3_2 = np.arange(17, 20)
moveAtomID_end_CH3_1 = np.array([0, 2, 3])
moveAtomID_end_CH3_2 = np.array([25, 26, 27])

delta_term_phi = np.pi * np.sign(phi_init) * phi_init / 180.0
delta_term_psi = np.pi * np.sign(psi_init) * psi_init / 180.0
delta_term_chi1 = np.pi * np.sign(chi1_init) * chi1_init / 180.0
delta_term_CH3_1 = np.pi * np.sign(CH3_1_init) * CH3_1_init / 180.0
delta_term_CH3_2 = np.pi * np.sign(CH3_2_init) * CH3_2_init / 180.0
delta_term_end_CH3_1 = np.pi * np.sign(end_CH3_1_init) * end_CH3_1_init / 180.0
delta_term_end_CH3_2 = np.pi * np.sign(end_CH3_2_init) * end_CH3_2_init / 180.0

# get clash list that include side chain CH3 atoms
ind0 = np.isin(clash_list[:, 0], moveAtomID_CH3_1)
ind1 = np.isin(clash_list[:, 1], moveAtomID_CH3_1)
ind2 = np.isin(clash_list[:, 0], moveAtomID_CH3_2)
ind3 = np.isin(clash_list[:, 1], moveAtomID_CH3_2)
CH3_clash_list = clash_list[ind0 | ind1 | ind2 | ind3, :]
CH3_radii_list = radii_2[ind0 | ind1 | ind2 | ind3]


# get clash list that includes end H3 atoms
ind0 = np.isin(clash_list[:, 0], moveAtomID_end_CH3_1)
ind1 = np.isin(clash_list[:, 1], moveAtomID_end_CH3_1)
CH3_end1_clash_list = clash_list[ind0 | ind1, :]
CH3_end1_radii_list = radii_2[ind0 | ind1]

ind0 = np.isin(clash_list[:, 0], moveAtomID_end_CH3_2)
ind1 = np.isin(clash_list[:, 1], moveAtomID_end_CH3_2)
CH3_end2_clash_list = clash_list[ind0 | ind1, :]
CH3_end2_radii_list = radii_2[ind0 | ind1]


# Get initial E
# Check clashes
diff_pos = Coord[clash_list[:, 0], :] - Coord[clash_list[:, 1], :]      # distance vector
sum_2 = np.sum(np.square(diff_pos), 1)                                  # distance^2
ind0 = sum_2 < radii_2                                                  # Compare the distance to the (sum_radii)^2
s_r_6 = np.power(radii_2[ind0] / sum_2[ind0], 3)
E = np.power(1 - s_r_6, 2)
total_E = np.sum(E)

min_E = 1000000


# All energy data
all_energy = np.zeros((72 * 72 * 72, 4))
counter = 0

# Store initial position so you can keep moving back
Initial_Position = Coord

# Loop over all phi
for phi_loop in range(0, 72):
    setPhi = phi_loop * 5.0
    Pos_phi = rotate_DA(Initial_Position.copy(), setPhi, delta_term_phi, phi_index, moveAtomID_phi)

    for psi_loop in range(0, 72):
        setPsi = psi_loop * 5.0
        Pos_phi_psi = rotate_DA(Pos_phi.copy(), setPsi, delta_term_psi, psi_index, moveAtomID_psi)

        for chi1_loop in range(0, 3):

            setChi = chi1_loop * 120.0 + 60.0
            Position_ppc = rotate_DA(Pos_phi_psi.copy(), setChi, delta_term_chi1, chi1_index, moveAtomID_chi1)
            total_E = 0

            # Now that we've rotated everything, check for clashes
            diff_pos = Position_ppc[clash_list[:, 0], :] - Position_ppc[clash_list[:, 1], :]
            sum_2 = np.sum(np.square(diff_pos), 1)
            ind0 = sum_2 < radii_2

            s_r_6 = np.power(radii_2[ind0] / sum_2[ind0], 3)
            E = np.power(1 - s_r_6, 2)
            total_E = np.sum(E)

            # See if any of the clashes included the SC CH3 group
            clash_used = clash_list[ind0, :]
            ind1 = np.isin(moveAtomID_CH3_1, clash_used)
            ind2 = np.isin(moveAtomID_CH3_2, clash_used)

            # If so, rotate the SC CH3 groups
            if (total_E > 0 and (np.any(ind1) or np.any(ind2))):

                min_E = total_E
                # Rotate CH3 groups
                [min_E, Position_CH3] = rotate_CH3(Position_ppc.copy(), delta_term_CH3_1, delta_term_CH3_2, CH3_1_index,
                                                   CH3_2_index, moveAtomID_CH3_1, moveAtomID_CH3_2, clash_list, radii_2,
                                                   min_E, CH3_clash_list, CH3_radii_list)
                total_E = min_E
                Position_ppc = Position_CH3.copy()

            # See if any clashes include N-terminal end CH3
            ind1 = np.isin(moveAtomID_end_CH3_1, clash_used)
            if (total_E > 0 and np.any(ind1)):
                min_E = total_E
                [min_E, new_Pos] = rotate_end_CH3(Position_ppc.copy(), delta_term_end_CH3_1, end_CH3_1_index,
                                                  moveAtomID_end_CH3_1, clash_list, radii_2, min_E,
                                                  CH3_end1_clash_list, CH3_end1_radii_list)
                total_E = min_E
                Position_ppc = new_Pos.copy()

            # See if any clashes include Cterminal end CH3
            ind2 = np.isin(moveAtomID_end_CH3_2, clash_used)
            if (total_E > 0 and np.any(ind2)):
                min_E = total_E
                [min_E, new_Pos] = rotate_end_CH3(Position_ppc.copy(), delta_term_end_CH3_2, end_CH3_2_index,
                                                  moveAtomID_end_CH3_2, clash_list, radii_2, min_E,
                                                  CH3_end2_clash_list, CH3_end2_radii_list)
                total_E = min_E
                Position_ppc = new_Pos.copy()

            all_energy[counter, 0] = setPhi
            all_energy[counter, 1] = setPsi
            all_energy[counter, 2] = setChi
            all_energy[counter, 3] = total_E
            counter = counter + 1


np.savetxt(save_folder + 'Val_energy_' + save_name + '_' + coord_num + '.txt', all_energy, fmt='%d %d %d %7.4f')

File: C_code/rotate_amino_acid.cxx
/******************************************************************************************************************************
 rotate_amino_acid.cxx

 Samples full phi/psi/chi space of an amino acid. Can change the step size of sampling

 Dependencies: dihedral_functions.cxx, dihedral_functions.h

 Compile: g++ rotate_amino_acid.cxx dihedral_functions.cxx

 Arguements:
 	1. folder:		String of the full path to the folder containing the input coordiantes file
 	2. input_file:	String of the file name of the input coordinate file
 	3. AA_name:		3 letter abbreviation of the amino acid
 	4. coord_number: integer of which set of coordiantes is being used (used in name of saved files)
 	5. save_folder:	full path to folder to save data
 	6. save_name:	additional string to used in the name of saved files (ex. 042221 to specify the date)
******************************************************************************************************************************/

#include "dihedral_functions.h" 
#include "rotate_CH3_group.h"
#include "aa_setup.h" 

#include <iostream>
#include <fstream>
#include <ostream>
#include <cstdlib>
#include <cmath>
#include <string.h>
#include <stdlib.h>     /* srand, rand */


using namespace std;
#define PI 3.14159265


int main(int argc, char **argv){

	/*******************************
	 * Process initial arguments
	 *******************************/
	if (argc < 6){
		cerr << " Invalid input " << endl ;
		cerr << endl;
		return 0;
	}
	
	
	string folder, input_file, AA_name, save_folder, save_name, coord_num;	

	folder = argv[1];		// Get folder name	
	input_file = argv[2];	// Get file name
	AA_name = argv[3];		// Get AA name
	coord_num = argv[4];	// Get which coord
	save_folder= argv[5];	// Get save folder
	save_name= argv[6];		// get same name

	// Initialize variabales
	int return_value = 0;
	int i, j, k;

	

	// Get dihedral angle indexes
	int index_array1[9][4];
	get_dihedral_indexes(AA_name, index_array1);

	int phi_index[4] = {index_array1[0][0], index_array1[0][1], index_array1[0][2], index_array1[0][3]};
	int psi_index[4] = {index_array1[1][0], index_array1[1][1], index_array1[1][2], index_array1[1][3]};
	int chi1_index[4] = {index_array1[2][0], index_array1[2][1], index_array1[2][2], index_array1[2][3]};
	int chi2_index[4] = {index_array1[3][0], index_array1[3][1], index_array1[3][2], index_array1[3][3]};
	int CH3_1_index[4] = {index_array1[5][0], index_array1[5][1], index_array1[5][2], index_array1[5][3]};
	int CH3_2_index[4] = {index_array1[6][0], index_array1[6][1], index_array1[6][2], index_array1[6][3]};
	int end_CH3_1_index[4] = {index_array1[7][0], index_array1[7][1], index_array1[7][2], index_array1[7][3]};
	int end_CH3_2_index[4] = {index_array1[8][0], index_array1[8][1], index_array1[8][2], index_array1[8][3]};


	// Get move atom ID indexes. First value is the actual lenght of the array
	int moveAtomID_phi[30], moveAtomID_psi[30], moveAtomID_chi1[30], moveAtomID_chi2[30], moveAtomID_chi3[30], moveAtomID_CH3_1[30], moveAtomID_CH3_2[30], moveAtomID_end_CH3_1[30], moveAtomID_end_CH3_2[30] ;

	get_move_indexes(AA_name, moveAtomID_phi, moveAtomID_psi, moveAtomID_chi1, moveAtomID_chi2, moveAtomID_chi3, moveAtomID_CH3_1, moveAtomID_CH3_2, moveAtomID_end_CH3_1, moveAtomID_end_CH3_2);

	int n_atoms, clash_size, CH3_clash_size, n_chi;
	if (AA_name.compare("Val") == 0){
		clash_size = 303;		// number of clashes (length of Clash_list_val.txt)
		CH3_clash_size = 129;	// number of clashes involving terminal CH3 atoms (length of CH3_clash_list_val.txt)
		n_atoms = 28;
		n_chi = 1;
	}
	else if (AA_name.compare("Ile") == 0){
		clash_size = 381;
		CH3_clash_size = 147;
		n_atoms = 31;
		n_chi = 2;
	}

	
	
	// Load clash list file
	int clash_atom1[clash_size], clash_atom2[clash_size]; 		// variables to hold atom indexes
	float radii_sum[clash_size], radii2[clash_size];		// variables to hold sum of radii and (sum of radii)^2
	string clash_file = "Clash_list_val.txt";
	string radii_file = "Clash_list_radii_sum_val.txt";
	load_clash_array(clash_file, clash_size, clash_atom1, clash_atom2);
	load_radii_array(radii_file, clash_size, radii_sum);
	
	//square the radii
	for (i = 0; i < clash_size; i ++){
	  		radii2[i] = radii_sum[i] * radii_sum[i];
	  	}

	// load CH3 clashlist file
	int CH3_clash_atom1[CH3_clash_size], CH3_clash_atom2[CH3_clash_size];		// varables to hold atom indexes
	float CH3_radii2[CH3_clash_size];											// holds (sum of radii)^2
	clash_file = "CH3_clash_list_val.txt";
	radii_file = "CH3_clash_list_radii2_sum_val.txt";
	load_clash_array(clash_file, CH3_clash_size, CH3_clash_atom1, CH3_clash_atom2);
	load_radii_array(radii_file, CH3_clash_size, CH3_radii2); // This radii is already squared

	// Load coordinates
	float Position[n_atoms][3];
	ifstream myfile;
	myfile.open((folder + input_file).c_str());
	if (myfile.is_open())
	  {
	  for (i = 0; i < n_atoms; i ++){
	  		myfile >> Position[i][0] >> Position[i][1] >> Position[i][2];
	  	}
	  	myfile.close();
	  }
	else{
		cout << "problem with position File" << endl;
		exit(EXIT_FAILURE);
	}


	// Get initial dihedral values for each dihedral
	float phi_init, psi_init, chi1_init, chi2_init, CH3_1_init, CH3_2_init, end_CH3_1_init, end_CH3_2_init;
	phi_init = calculate_DA(phi_index, Position);
	psi_init= calculate_DA(psi_index, Position);
	chi1_init = calculate_DA(chi1_index, Position);
	chi2_init = calculate_DA(chi2_index, Position);
	CH3_1_init= calculate_DA(CH3_1_index, Position);
	CH3_2_init= calculate_DA(CH3_2_index, Position);
	end_CH3_1_init= calculate_DA(end_CH3_1_index, Position);
	end_CH3_2_init= calculate_DA(end_CH3_2_index, Position);

	// Delta term is used when rotating the dihedral angles. Since it is a constant we can calculate it once here
	float delta_term_phi, delta_term_psi, delta_term_chi1, delta_term_chi2, delta_term_CH3_1, delta_term_CH3_2, delta_term_end_CH3_1, delta_term_end_CH3_2;
	delta_term_phi = abs(PI * phi_init / 180.0);
	delta_term_psi = abs(PI *  psi_init / 180.0);
	delta_term_chi1 = abs(PI * chi1_init / 180.0);
	delta_term_chi2 = abs(PI * chi2_init / 180.0);
	delta_term_CH3_1 = abs(PI *  CH3_1_init / 180.0);
	delta_term_CH3_2 = abs(PI *  CH3_2_init / 180.0);
	delta_term_end_CH3_1 = abs(PI * end_CH3_1_init / 180.0);
	delta_term_end_CH3_2 = abs(PI * end_CH3_2_init / 180.0);

	// Declare a bunch of variables
	float newPos[n_atoms][3];
	float setPhi, setPsi, setChi1, setChi2, setCH3_1, setCH3_2, E, minE, CH3_E;
	float Pos_phi[n_atoms][3], Pos_psi[n_atoms][3], Pos_chi1[n_atoms][3], Pos_chi2[n_atoms][3], Pos_CH3_1[n_atoms][3], Pos_CH3_2[n_atoms][3];
	float CH3_min[n_atoms][3];
	float save_data[72 * 72][n_chi+3];
	int save_counter = 0;

	//Clear output file
	string n1, n2, n3, f1;
	n1 = "_energy_";
	n2 = "_";
	n3 = ".txt";
	f1 = save_folder + AA_name + n1 + save_name + n2 + coord_num + n3;
	ofstream out_file1(f1.c_str(), ios::out);
	out_file1 << fixed;

	for (int phi_loop = 0; phi_loop < 72; phi_loop++){					// loop over all phi values

		// Rotate to the new phi value
		setPhi = phi_loop * 5.0;
		rotate_DA(Position, setPhi, delta_term_phi, phi_index, moveAtomID_phi, n_atoms, Pos_phi);

		for (int psi_loop = 0; psi_loop < 72; psi_loop++){				// loop over all psi values

			// Rotate to the new psi valu
			setPsi = psi_loop * 5.0;
			rotate_DA(Pos_phi, setPsi, delta_term_psi, psi_index, moveAtomID_psi, n_atoms, Pos_psi);

			for (int chi1_loop =0; chi1_loop < 72; chi1_loop++ ){ 		// loop over all chi1 values

				// Rotate to the new chi1 valu
				setChi1 = chi1_loop * 5.0;
				rotate_DA(Pos_psi, setChi1, delta_term_chi1, chi1_index, moveAtomID_chi1, n_atoms, Pos_chi1);
				if (n_chi >= 2){

					for (int chi2_loop = 0; chi2_loop < 72; chi2_loop++){
					
						setChi2 = chi2_loop * 5.0;
						rotate_DA(Pos_chi1, setChi2, delta_term_chi2, chi2_index, moveAtomID_chi2, n_atoms, Pos_chi2);
						// Get energy
						E = get_energy(Pos_chi2, radii2, clash_atom1, clash_atom2, clash_size);

						// now we have to consider the terminal CH3 groups
						// Get energy due to CH3 atoms
						CH3_E = get_energy(Pos_chi2, CH3_radii2, CH3_clash_atom1, CH3_clash_atom2, CH3_clash_size);
						// If there is energy due to overlap of the CH3 hydrogens, rotate these extra dihedrals, finding the lowest energy configuration
						if (CH3_E > 0){	

							rotate_CH3_group(n_atoms, Pos_chi2, CH3_radii2, CH3_clash_atom1, CH3_clash_atom2, CH3_clash_size, delta_term_CH3_1, delta_term_CH3_2, CH3_1_index, CH3_2_index, moveAtomID_CH3_1, moveAtomID_CH3_2, CH3_min);
							E = get_energy(CH3_min, radii2, clash_atom1, clash_atom2, clash_size);
						}

						// Save data
						save_data[save_counter][0] = setPhi;
						save_data[save_counter][1] = setPsi;
						save_data[save_counter][2] = setChi1;
						save_data[save_counter][3] = setChi2;
						save_data[save_counter][5] = E;
						save_counter = save_counter + 1;
						if (save_counter >= 72 * 72){
							for (i=0; i < save_counter; i ++){
								out_file1 << save_data[i][0] << " " << save_data[i][1] << " " << save_data[i][2] << " " << save_data[i][3] << " " << save_data[i][4] << endl;
							}
							save_counter = 0;
						}
					} // end chi2 loop
				} // end if Chi 2
				else {			// if only chi1
					// Get energy
					E = get_energy(Pos_chi1, radii2, clash_atom1, clash_atom2, clash_size);

					// now we have to consider the terminal CH3 groups
					// Get energy due to CH3 atoms
					CH3_E = get_energy(Pos_chi1, CH3_radii2, CH3_clash_atom1, CH3_clash_atom2, CH3_clash_size);
					// If there is energy due to overlap of the CH3 hydrogens, rotate these extra dihedrals, finding the lowest energy configuration
					if (CH3_E > 0){
						rotate_CH3_group(n_atoms, Pos_chi1, CH3_radii2, CH3_clash_atom1, CH3_clash_atom2, CH3_clash_size, delta_term_CH3_1, delta_term_CH3_2, CH3_1_index, CH3_2_index, moveAtomID_CH3_1, moveAtomID_CH3_2, CH3_min);
						E = get_energy(CH3_min, radii2, clash_atom1, clash_atom2, clash_size);
					}
					// Save data
					save_data[save_counter][0] = setPhi;
					save_data[save_counter][1] = setPsi;
					save_data[save_counter][2] = setChi1;
					save_data[save_counter][3] = E;
					save_counter = save_counter + 1;
					if (save_counter >= 72 * 72){
						for (i=0; i < save_counter; i ++){
							out_file1 << save_data[i][0] << " " << save_data[i][1] << " " << save_data[i][2] << " " << save_data[i][3] << endl;
						}
						save_counter = 0;
					}


				} // end of if only Chi 1
			} // end of Chi1 loop
		} // end psi loop
	} // end phi loop

	out_file1.close();
}